kazoo/protocol/connection.py

```diff
@@ -725,9 +727,7 @@ def _connect(self, host, hostip, port):
             client.read_only,
         )
 
-        connect_result, zxid = self._invoke(
-            client._session_timeout / 1000.0 / len(client.hosts), connect
-        )
+        connect_result, zxid = self._invoke(timeout, connect)
```

Comment:
I agree that `_session_timeout`, being the *requested* session timeout and thus potentially arbitrarily larger than the actual session timeout, is an unfortunate choice.

Thinking about it some more, I am not sure that using the retry's `cur_delay` as-is is a good choice either.  It may have been sufficient for the TCP (and potentially SSL) handshakes, but this request/response cycle involves a round trip through the JVM, disk logging, quorum packets (learner with global sessions), etc.

I can imagine many "false starts" (and thus gratuitous network & server activity) if the default `cur_delay` is kept too low.

Reply:
Interestingly, `_authenticate_with_sasl` and `AUTH_XID` below reuse `connect_timeout`—which, at least, *is* related to the negotiated timeout—for "numerous" write and read operations which have to complete timely as they do not "touch" the session.

The good news is that the session is "fresh" at that point.  But in theory, the negotiated timeout should be used to determine a *deadline for the whole sequence*, so that it aborts  before causing a session expiry if e.g. SASL authentication proceeds too slowly.

Reply:
(Note that this is not handled in the other clients either.)

Reply:
I think that the "session timeout" is an indication what you want (as you point out, before negotiation with the server's config) for how long your session will remain "alive" in case the client disconnects. In other words, how much time the client has to reconnect.

I have a more "radical" view than yours, I think, which is that it has *nothing* to do with how long it takes to establish TCP connection between the client and the servers.

The increasing backoff method is a good way to adapt to Zookeeper servers that can be deployed on the local network (~10s ms), across data centers (~100s ms) or across continents. The default value would be more fitting for the 2 first use cases and, yes, possibly less so for wider deployments, but anyway for such setups, it is possible to easily tweak that value.

That is my opinion for the *TCP* connection timeout.

For the requests timeout, which is once the TCP connection is established, this is something else.

Reply:
For request/reply messages on an established TCP connection, including `Connect` requests.

The original code used `client._session_timeout / 1000.0 / len(client.hosts)` for the Connect (pre-negotiation).

After that, all other request/reply use these two, `connect_timeout` for writes, `read_timeout` for reads:

``` python
connect_timeout = negotiated_session_timeout / len(client.hosts)
read_timeout = negotiated_session_timeout * 2.0 / 3.0
```

My rational for changing the timeout on the 1st `Connect` is that it is possible to have a healthy host (which succeeds to TCP connect) but an unhealthy JVM/Zookeeper that does not handle the `Connect` request.

In such case, the client would wait a long time before timing out and moving to the next server, much of its total session timeout.
It feels that waiting arbitrary long when we know what was the timeout of a successful TCP is not ideal, especially when we have not yet selected a server and there could be others. 

But I could be convinced that this should be part of a different change.